src/rgw/rgw_sal.h

```diff
@@ -1366,45 +1281,73 @@ class Lifecycle {
 public:
   /** Head of a lifecycle run.  Used for tracking parallel lifecycle runs. */
   struct LCHead {
-    time_t start_date{0};
-    std::string marker;
-    time_t shard_rollover_date{0};
-
     LCHead() = default;
-    LCHead(time_t _date, std::string& _marker) : start_date(_date), marker(_marker) {}
+    virtual ~LCHead() = default;
+
+    virtual time_t& get_start_date() = 0;
+    virtual void set_start_date(time_t) = 0;
+    virtual std::string& get_marker() = 0;
+    virtual void set_marker(const std::string&) = 0;
+    virtual time_t& get_shard_rollover_date() = 0;
+    virtual void set_shard_rollover_date(time_t) = 0;
   };
 
   /** Single entry in a lifecycle run.  Multiple entries can exist processing different
    * buckets. */
   struct LCEntry {
-    std::string bucket;
-    std::string oid;
-    uint64_t start_time{0};
-    uint32_t status{0};
-
     LCEntry() = default;
-    LCEntry(std::string& _bucket, uint64_t _time, uint32_t _status) : bucket(_bucket), start_time(_time), status(_status) {}
-    LCEntry(std::string& _bucket, std::string _oid, uint64_t _time, uint32_t _status) : bucket(_bucket), oid(_oid), start_time(_time), status(_status) {}
+    virtual ~LCEntry() = default;
+
+    virtual std::string& get_bucket() = 0;
+    virtual void set_bucket(const std::string&) = 0;
+    virtual std::string& get_oid() = 0;
+    virtual void set_oid(const std::string&) = 0;
+    virtual uint64_t get_start_time() = 0;
+    virtual void set_start_time(uint64_t) = 0;
+    virtual uint32_t get_status() = 0;
+    virtual void set_status(uint32_t) = 0;
```

Comment:
@dang can you remind me why `struct LCHead` and `LCEntry` had to change into virtual base classes? the pr says this was a "prereq for filters"

filters do need the ability to wrap abstractions, but do we really want abstractions for LCHead and LCEntry? aren't they just data? `rgw::sal::Lifecycle` provides the backend abstraction

it seems unnecessary for `Lifecycle::list_entries()` to be allocating a `unique_ptr<LCEntry>` for each entry it returns, and then for filters to allocate another `FilterLCEntry` to wrap each of those. would you object to my changing back to raw structs?

cc @mattbenjamin

Reply:
I don't remember for sure, but my guess is that you can't have a non-pure-virtual class inside a pure-virtual class.  So they could be moved outside of Lifecycle and turned into structs, I guess.

Reply:
thanks, i pushed a draft to https://github.com/ceph/ceph/pull/58603 but it's based on other changes for now